tests/src/python/test_qgsvectorlayerutils.py

```diff
@@ -690,6 +690,49 @@ def test_unique_pk_when_subset(self):
         vl.addFeatures(features)
         self.assertTrue(vl.commitChanges())
 
+    def test_check_attribute_type(self):
+        """Test checkAttributeType"""
+
+        vl = QgsVectorLayer('Point?crs=epsg:4326&field=int:integer', 'test', 'memory')
+
+        # Valid values
+        self.assertTrue(QgsVectorLayerUtils.canConvert(123.0, vl.fields()[0].type()))
+        self.assertTrue(QgsVectorLayerUtils.canConvert(123, vl.fields()[0].type()))
+        # Check NULL/invalid
+        self.assertTrue(QgsVectorLayerUtils.canConvert(None, vl.fields()[0].type()))
+        self.assertTrue(QgsVectorLayerUtils.canConvert(QVariant(QVariant.Int), vl.fields()[0].type()))
+        # Not valid
+        self.assertFalse(QgsVectorLayerUtils.canConvert('QGIS Rocks!', vl.fields()[0].type()))
+        self.assertFalse(QgsVectorLayerUtils.canConvert(QDate(2020, 6, 30), vl.fields()[0].type()))
+        # Not valid: overflow and narrow cast!
+        self.assertFalse(QgsVectorLayerUtils.canConvert(2147483647 + 1, vl.fields()[0].type()))
+        self.assertFalse(QgsVectorLayerUtils.canConvert(123.123, vl.fields()[0].type()))
+
+        vl = QgsVectorLayer('Point?crs=epsg:4326&field=date:date', 'test', 'memory')
+        self.assertTrue(QgsVectorLayerUtils.canConvert(QDate(2020, 6, 30), vl.fields()[0].type()))
+        # Not valid
+        self.assertFalse(QgsVectorLayerUtils.canConvert('QGIS Rocks!', vl.fields()[0].type()))
```

Comment:
needs a test for date string values, e.g. '2020-01-03', '30/01/2020',... (these should be accepted)